feat: [CDS-91799]: Add reconciliation concept in the doc (#6072)
harness/developer-hub: docs/platform/templates/reconcile-pipeline-templates.md

sidebar_position: 15
---

Harness detects updates made to referenced entities if those updates are made outside of Pipeline Studio. When you are viewing a pipeline in Pipeline Studio, if a referenced entity has an update, Harness informs you and prompts you to update the pipeline. This process of updating the referenced entities in the pipeline is called pipeline reconciliation. The process ensures that you are aware that referenced entities have updates, and you can choose to integrate those changes into the pipeline. 

> One of the most example of pipeline reconciliation is when you increase or decrease the number of runtime inputs.

Reviewer: most **common** example

sidebar_position: 15
---

> Harness detects updates made to referenced entities if those updates are made outside or inside of the Pipeline Studio. When you are viewing a pipeline in Pipeline Studio, if a referenced entity has an update, Harness informs you and prompts you to update the pipeline. This process of updating the referenced entities in the pipeline is called pipeline reconciliation. The process ensures that you are aware that referenced entities have updates, and you can choose to integrate those changes into the pipeline. 

Reviewer: Can we define what a "referenced entity" is before jumping into this?

Author: @SudarshanPurohit already defined in the first line along with the doc link.

> Harness detects updates made to referenced entities if those updates are made outside or inside of the Pipeline Studio. When you are viewing a pipeline in Pipeline Studio, if a referenced entity has an update, Harness informs you and prompts you to update the pipeline. This process of updating the referenced entities in the pipeline is called pipeline reconciliation. The process ensures that you are aware that referenced entities have updates, and you can choose to integrate those changes into the pipeline. 

Reviewer 2: ```suggestion
If any [entity](https://developer.harness.io/docs/platform/references/harness-entity-reference/) referenced in a pipeline is updated, Harness prompts you to update the pipeline if you're viewing the pipeline in Pipeline Studio. Harness detects updates made to the referenced entities using Pipeline Studio or YAML view. This process of updating the referenced entities in the pipeline is called pipeline reconciliation. The process ensures that you are aware that referenced entities have updates, and you can choose to integrate those changes into the pipeline. 
```

> One of the most common example of pipeline reconciliation is when you increase or decrease the number of runtime inputs.

Reviewer: Why does a change in runtime inputs require reconciliation? We could add it here, or else point to it if already defined elsewhere.

Author: defined below with an example

> For example, let's consider a scenario with three templates: a pipeline template, a stage template, and a step template. In this scenario, the pipeline serves as the parent template, while the stage and step templates are its child templates. 

Reviewer: But the example below doesn't have the step template as a child of the pipeline template? it is a child of the stage template instead.

Author: Corrected the same

```

**Pipeline Template**
```yaml
template:
  name: pipeline_template_1
  identifier: pipeline_template_1
  versionLabel: v1
  type: Pipeline
  projectIdentifier: Pipeline_Samples
  orgIdentifier: default
  tags: {}
  spec:
    stages:
      - stage:
          name: stagetemp1
          identifier: stagetemp1
          tags: {}
          template:
            templateRef: stage_template_1
            versionLabel: v1

```
In this pipeline template, both the step as well stage template are utilized. Now, when you update a variable at a step level for example, you have changed the value of variable from ``Fixed value`` to ``Runtime input``.

**Step template**
```yaml
template:
  name: step_template_1
  identifier: step_template_1
  versionLabel: v1
  type: Step
  projectIdentifier: Pipeline_Samples
  orgIdentifier: default
  tags: {}
  spec:
    timeout: 10m
    type: ShellScript
    spec:
      shell: Bash
      executionTarget: {}
      delegateSelectors: []
      source:
        type: Inline
        spec:
          script: echo hello
      environmentVariables:
        - name: var1
          type: String
          value: <+input>
      outputVariables: []
```

> When you have updated a child entity, in this case, the step template, the parent entities, i.e., stage and pipeline templates, need to be updated as well. In this scenario, the concept of reconciliation comes into play.

Reviewer: maybe add a line to explain the concept you mentioned before - that the number of runtime inputs has decreased from A to B. So, everyone who's referring to that template now needs to provide B inputs instead of A.

Author: @SudarshanPurohit in this example I explained that when you changed the input type from Fixed value to Runtime Input then also reconciliation comes into picture

Now, when you move to the pipeline template, you will see a message: ``Some of the entities referenced in this template have gone out of sync.`` along with a ``Reconcile`` button.
![](./static/reconcile_popup.png)

> Harness calls the refreshed YAML API `POST https://app.harness.io/template/api/refresh-template/refreshed-yaml` when you select the **Reconcile** option in the **Pipeline Studio**. This API gets the latest pipeline YAML. Harness shows the differences between the original and the refreshed YAML in the UI on the Template Error Inspection page.

Reviewer: This is good data. It's probably worth writing out how a reconciliation would be done purely through APIs. @utkarshchoubey-harness should be able to help here.

![](./static/reconcile_pipeline_template.png)

First, it will ask you to update the stage template, and then the pipeline level. You also have the option to ``Update all unsynced entities`` at once.
:::info note
Please note that every reconciliation requires saving the pipeline. Specifically, if the pipeline is remote, we will need to do a ``git push`` every time to ensure the latest version of the pipeline is saved.
:::

> Pipeline reconciliation allows you to add or remove ``allowed values``. However if you add new values to the parent entity it will not allow the same. Parent entity can have subset of values that are present in child entity.

Reviewer: Add or remove allowed values to what? and where?

For example, You have a Pipeline that uses a stage and step template. 

Step template
```yaml
template:
  name: step_template_1
  identifier: step_template_1
  versionLabel: v1
  type: Step
  projectIdentifier: CD_Samples
  orgIdentifier: default
  tags: {}
  spec:
    timeout: 10m
    type: ShellScript
    spec:
      shell: Bash
      executionTarget: {}
      delegateSelectors: []
      source:
        type: Inline
        spec:
          script: echo hello
      environmentVariables:
        - name: var1
          type: String
          value: <+input>.allowedValues(a,b,c,d)
      outputVariables: []

```

> Now, the parent entity i.e stage or pipeline can have subsets of the above mentioned ``allowed values``. But if you add a new value for example ``e`` then it won't allow you to add the same. 

Reviewer: why won't it? better to explain

sidebar_position: 15
---

If any [entity](https://developer.harness.io/docs/platform/references/harness-entity-reference/) referenced in a pipeline is updated, Harness prompts you to update the pipeline if you're viewing the pipeline in Pipeline Studio. Harness detects updates made to the referenced entities using Pipeline Studio or YAML view. 

> This process of updating the referenced entities in the pipeline is called pipeline reconciliation. The process ensures that you are aware that referenced entities have updates, and you can choose to integrate those changes into the pipeline. 

Reviewer 2: ```suggestion
This process of updating the referenced entities in the pipeline is called pipeline reconciliation. It ensures that you are aware that referenced entities have updates and you can choose to integrate those changes into the pipeline. 
```

One of the most common example of pipeline reconciliation is when you increase or decrease the number of runtime inputs.

> For example, let's consider a scenario with three templates: a pipeline template, a stage template, and a step template. In this scenario, the pipeline serves as the parent template, while the stage is its child template, and the step serves as the stage's child template

Reviewer 2: ```suggestion
Let's consider a scenario with three templates: a pipeline template, a stage template, and a step template. In this scenario, the pipeline serves as the parent template, the stage is the pipeline's child template, and the step is the stage's child template.
```

```

**Pipeline Template**
```yaml
template:
  name: pipeline_template_1
  identifier: pipeline_template_1
  versionLabel: v1
  type: Pipeline
  projectIdentifier: Pipeline_Samples
  orgIdentifier: default
  tags: {}
  spec:
    stages:
      - stage:
          name: stagetemp1
          identifier: stagetemp1
          tags: {}
          template:
            templateRef: stage_template_1
            versionLabel: v1

```
In this pipeline template, both the step as well stage template are utilized. Now, when you update a variable at a step level for example, you have changed the value of variable from ``Fixed value`` to ``Runtime input``.

**Step template**
```yaml
template:
  name: step_template_1
  identifier: step_template_1
  versionLabel: v1
  type: Step
  projectIdentifier: Pipeline_Samples
  orgIdentifier: default
  tags: {}
  spec:
    timeout: 10m
    type: ShellScript
    spec:
      shell: Bash
      executionTarget: {}
      delegateSelectors: []
      source:
        type: Inline
        spec:
          script: echo hello
      environmentVariables:
        - name: var1
          type: String
          value: <+input>
      outputVariables: []
```

> When you have updated a child entity, in this case, the step template, the parent entities, i.e., stage and pipeline templates, need to be updated as well. In this scenario, the concept of reconciliation comes into play. 

Reviewer 2: ```suggestion
When you update a child entity, the step template and the parent entities, i.e., stage and pipeline templates must be updated as well. In this scenario, the concept of reconciliation comes into play. 
```

> Now, when you move to the pipeline template, you will see a message: ``Some of the entities referenced in this template have gone out of sync.`` along with a ``Reconcile`` button.

Reviewer 2: ```suggestion

When you move to the pipeline template, you can see the message: ``Some of the entities referenced in this template have gone out of sync.`` along with a ``Reconcile`` button.
```

![](./static/reconcile_popup.png)

> Harness calls the refreshed YAML API `POST https://app.harness.io/template/api/refresh-template/refreshed-yaml` when you select the **Reconcile** option in the **Pipeline Studio**. This API gets the latest pipeline YAML. Harness shows the differences between the original and the refreshed YAML in the UI on the Template Error Inspection page.

Reviewer 2: ```suggestion
Harness calls the refreshed YAML API `POST https://app.harness.io/template/api/refresh-template/refreshed-yaml` when you select the **Reconcile** option in **Pipeline Studio**. This API gets the latest pipeline YAML. Harness shows the difference between the original and the refreshed YAML in the UI on the Template Error Inspection page.
```

![](./static/reconcile_pipeline_template.png)

> First, it will ask you to update the stage template, and then the pipeline level. You also have the option to ``Update all unsynced entities`` at once.

Reviewer 2: ```suggestion
First, it will ask you to update the stage template, and then the pipeline template level. You also have the option to ``Update all unsynced entities`` at once.
```

:::info note

> Please note that every reconciliation requires saving the pipeline. Specifically, if the pipeline is remote, we will need to do a ``git push`` every time to ensure the latest version of the pipeline is saved.

Reviewer 2: ```suggestion
Every reconciliation requires saving the pipeline. Specifically, if the pipeline is remote, you must to do a ``git push`` every time to ensure that the latest version of the pipeline is saved.
```

:::

> Pipeline reconciliation allows you to add or remove ``allowed values`` to your runtime inputs. However if you add new allowed values to the parent entity it will not allow the same. Parent entity can have subset of allowed values that are present in child entity.

Reviewer 2: ```suggestion
Pipeline reconciliation allows you to add or remove ``allowed values`` to your runtime inputs. However, if you add new allowed values to the parent entity, it will not allow the same. Parent entity can have subset of allowed values that are present in child entity.
```

> For example, You have a Pipeline that uses a stage and step template. 

Reviewer 2: ```suggestion
For example, You have a pipeline that uses a stage and step template. 
```

Step template
```yaml
template:
  name: step_template_1
  identifier: step_template_1
  versionLabel: v1
  type: Step
  projectIdentifier: CD_Samples
  orgIdentifier: default
  tags: {}
  spec:
    timeout: 10m
    type: ShellScript
    spec:
      shell: Bash
      executionTarget: {}
      delegateSelectors: []
      source:
        type: Inline
        spec:
          script: echo hello
      environmentVariables:
        - name: var1
          type: String
          value: <+input>.allowedValues(a,b,c,d)
      outputVariables: []

```

> Now, the parent entity i.e stage or pipeline can have subsets of the above mentioned ``allowed values``. But if you add a new value for example ``e`` in the parent entity then it won't allow you to add the same since child entity in this case step template doesn't have that updated value in it's allowed value and the parent entity in the above scenario is picking up allowed values from child entity. 

Reviewer 2: ```suggestion
The parent entities like stage or pipeline can have subsets of the above mentioned ``allowed values``. But, if you add a new value, for example,``e`` in the parent entity, then it won't allow you to add the same since the child entity in the step template doesn't have that updated value in it's allowed value. The parent entity in the above scenario picks up allowed values from the child entity. 
```

- A field's value type was changed from a fixed value or expression to a runtime input, or from a runtime input to a fixed value or expression.

:::info note

> Please note, reconciliation doesn't allow you to add or remove allowed values. Suppose you add or remove a value to allowed values; reconciliation will remove it.

Reviewer: Actually reconciliation should allow you to *remove* allowed values, but not add. That's because we can restrict some of the template's allowed values from the pipeline, but not add more allowed values (the template may not be expecting those). Please check with @sunnykesh-harness and @PratyushGar on this.

Author: Hey @SudarshanPurohit yes I checked just now. Last discussed with @utkarshchoubey-harness it was not possible to remove the values. I guess it's fixed. @utkarshchoubey-harness can you please confirm on this?
